edelphi/src/main/java/fi/metatavu/edelphi/pages/panel/admin/report/thesis/AbstractThesisScale2DQueryReportPage.java

```diff
@@ -0,0 +1,177 @@
+
+      if (answerX != null && answerY != null) {
+        int x = NumberUtils.createInteger(answerX.getOption().getValue());
+        int y = NumberUtils.createInteger(answerY.getOption().getValue());
+
+        values[x][y] = values[x][y] != null ? values[x][y] + 1 : 1; 
+      }
+    }
+
+    QueryPageSettingDAO queryPageSettingDAO = new QueryPageSettingDAO();
+    QueryPageSettingKeyDAO queryPageSettingKeyDAO = new QueryPageSettingKeyDAO();
+    QueryPageSettingKey queryPageSettingKey = queryPageSettingKeyDAO.findByName("scale2d.label.x");
+    QueryPageSetting queryPageSetting = queryPageSettingDAO.findByKeyAndQueryPage(queryPageSettingKey, queryPage);
+    String xLabel = queryPageSetting == null ? null : queryPageSetting.getValue();
+    queryPageSettingKey = queryPageSettingKeyDAO.findByName("scale2d.label.y");
+    queryPageSetting = queryPageSettingDAO.findByKeyAndQueryPage(queryPageSettingKey, queryPage);
+    String yLabel = queryPageSetting == null ? null : queryPageSetting.getValue();
+
+    return ChartModelProvider.createBubbleChart(title, xLabel, xTickLabels, yLabel, yTickLabels, 0, 0, values);
+  }
+
+  protected Chart createBarChart(ChartContext chartContext, QueryPage queryPage, String title, Render2dAxis render2dAxis, String fieldName) {
+    QueryFieldDAO queryFieldDAO = new QueryFieldDAO();
+    QueryOptionFieldOptionDAO queryOptionFieldOptionDAO = new QueryOptionFieldOptionDAO();
+    
+    QueryOptionField queryField = (QueryOptionField) queryFieldDAO.findByQueryPageAndName(queryPage, fieldName);
+    List<QueryOptionFieldOption> queryFieldOptions = queryOptionFieldOptionDAO.listByQueryField(queryField);
+
+    List<QueryReply> queryReplies = ReportUtils.getQueryReplies(queryPage, chartContext.getReportContext());
+    Map<Long, Long> data = ReportUtils.getOptionListData(queryField, queryFieldOptions, queryReplies);
+    
+    List<Double> values = new ArrayList<>();
+    List<String> categoryCaptions = new ArrayList<>();
+    
+    for (QueryOptionFieldOption optionFieldOption : queryFieldOptions) {
+      Long optionId = optionFieldOption.getId();
+      categoryCaptions.add(optionFieldOption.getText());
+      values.add(Double.valueOf(data.get(optionId)));
+    }
+    
+    // Axis label
+    
+    QueryPageSettingDAO queryPageSettingDAO = new QueryPageSettingDAO();
+    QueryPageSettingKeyDAO queryPageSettingKeyDAO = new QueryPageSettingKeyDAO();
+    String labelSettingName = render2dAxis == Render2dAxis.X ? "scale2d.label.x" : "scale2d.label.y";
+    QueryPageSettingKey queryPageSettingKey = queryPageSettingKeyDAO.findByName(labelSettingName);
+    QueryPageSetting queryPageSetting = queryPageSettingDAO.findByKeyAndQueryPage(queryPageSettingKey, queryPage);
+    String labelText = queryPageSetting == null ? null : queryPageSetting.getValue();
+
+    // Statistics
+    // TODO These could be calculated elsewhere and added below the chart image?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1135)